src/boot/efi/stub.c

```diff
@@ -360,6 +360,137 @@ static void dt_filenames_free(char16_t **dt_filenames, size_t n_dt) {
+                void **ret_ucode_base,
+                size_t *ret_ucode_size,
+                int *ret_sections_measured) {
+
+        int sections_measured = -1;
+        EFI_STATUS err;
+
+        assert(ret_ucode_base);
+        assert(ret_ucode_size);
+        assert(ret_sections_measured);
+
+        _cleanup_pages_ Pages global_addons_combined_pages = {}, uki_addons_combined_pages = {}, all_ucode_combined_pages = {};
+        void *global_addons_combined_base = NULL, *uki_addons_combined_base = NULL;
+        size_t global_addons_combined_size = 0, uki_addons_combined_size = 0, all_ucode_combined_size = 0;
+
+        if (n_ucode_addons_global == 0 && n_ucode_addons_uki == 0 && ucode_section_size == 0)
+                return EFI_SUCCESS;
+
+        /* Order of ucode merging (and measurements):
+         *  1. UKI addons
+         *  2. Global addons
+         *  3. UKI embedded .ucode section
+         */
+
+        /* Measure UKI addons */
+        for (size_t i = 0; i < n_ucode_addons_uki; i++)
+                measure_uki_section(POINTER_TO_PHYSICAL_ADDRESS(ucode_bases_addon_uki[i]), ucode_sizes_addon_uki[i], UNIFIED_SECTION_UCODE, &sections_measured);
+
+        /* Merge UKI addons */
+        if (n_ucode_addons_uki == 1) {
+                uki_addons_combined_base = ucode_bases_addon_uki[0];
+                uki_addons_combined_size = ucode_sizes_addon_uki[0];
+        } else if (n_ucode_addons_uki > 1) {
+                err = combine_initrds(*ucode_bases_addon_uki, ucode_sizes_addon_uki, n_ucode_addons_uki, &uki_addons_combined_pages, &uki_addons_combined_size);
+                if (err != EFI_SUCCESS)
+                        return err;
+                uki_addons_combined_base = PHYSICAL_ADDRESS_TO_POINTER(uki_addons_combined_pages.addr);
+        }
+
+        /* Measure global addons */
+        for (size_t i = 0; i < n_ucode_addons_global; i++)
+                measure_uki_section(POINTER_TO_PHYSICAL_ADDRESS(ucode_bases_addon_global[i]), ucode_sizes_addon_global[i], UNIFIED_SECTION_UCODE, &sections_measured);
+
+        /* Merge global addons */
+        if (n_ucode_addons_global == 1) {
+                global_addons_combined_base = ucode_bases_addon_global[0];
+                global_addons_combined_size = ucode_sizes_addon_global[0];
+        } else if (n_ucode_addons_global > 1) {
+                err = combine_initrds(*ucode_bases_addon_global, ucode_sizes_addon_global, n_ucode_addons_global, &global_addons_combined_pages, &global_addons_combined_size);
+                if (err != EFI_SUCCESS)
+                        return err;
+                global_addons_combined_base = PHYSICAL_ADDRESS_TO_POINTER(global_addons_combined_pages.addr);
+        }
+
+        /* Measure .ucode section from UKI again */
+        if (ucode_section_size > 0)
+                measure_uki_section(POINTER_TO_PHYSICAL_ADDRESS(ucode_section_base), ucode_section_size, UNIFIED_SECTION_UCODE, &sections_measured);
```

Comment:
This was important before when I excluded the UKI embedded `.ucode` section from the initial measurement. But now it is measured there - does measuring it again here give us anything? Ucode is tiny, it shouldn't really cost anything, opinions on keep vs kick out?